File: 18. 4Sum.py
# timeout

class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        n=len(nums)
        nums=sorted(nums)
        res=[]
        i=0
        while i < n:
            j=i+1
            while j < n:
                k=j+1
                l=n-1
                while l>k:
                    if nums[i]+nums[k]+nums[j]+nums[l] == target:
                        res+=[[nums[i], nums[j], nums[k], nums[l]]]
                        k+=1
                        l-=1
                        while k<j and nums[k]==nums[k-1]:
                            k+=1
                        while l>j and l<n-1 and nums[l]==nums[l+1]:
                            l-=1
                    elif nums[i]+nums[k]+nums[j]+nums[l] > target:
                        l-=1
                        while l>j and l<n-1 and nums[l]==nums[l+1]:
                            l-=1
                    else:
                        k+=1
                        while k<j and nums[k]==nums[k-1]:
                            k+=1
                j+=1
                while j<n and nums[j]==nums[j-1]:
                    j+=1
            i+=1
            while i<n and nums[i]==nums[i-1]:
                i+=1
        return res

File: 18. 4Sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        
        int i=0;
        while (i < nums.size())
        {
            int j = i + 1;
            while (j < nums.size())
            {
                int k = j+1;
                int l = nums.size()-1;
                while (k < l)
                {
                    if (nums[i]+nums[j]+nums[k]+nums[l] == target)
                    {
                        vector<int> temp {nums[i], nums[j], nums[k], nums[l]};
                        res.push_back(temp);
                        k++;
                        l--;
                        while (k < nums.size() && nums[k] == nums[k-1])
                            k++;
                        while (l > j && nums[l] == nums[l+1])
                            l--;
                    }
                    else if (nums[i]+nums[j]+nums[k]+nums[l] > target)
                    {
                        l--;
                        while (l > j && nums[l] == nums[l+1])
                            l--;
                    }
                    else
                    {
                        k++;
                        while (k < nums.size() && nums[k] == nums[k-1])
                            k++;
                    }
                }
                j++;
                while (j < nums.size() and nums[j] == nums[j-1])
                    j++;
            }
            i++;
            while (i < nums.size() and nums[i] == nums[i-1])
                i++;
        }
        return res;
    }
};
